delta/rdf/src/main/scala/ch/epfl/bluebrain/nexus/delta/rdf/shacl/ShaclEngine.scala

```diff
@@ -0,0 +1,108 @@
+
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Extend the [[ValidationEngine]] form TopQuadrant in order to add triples to the report
+  * with the number of targetedNodes
+  *
+  * @param dataset        the Dataset to operate on
+  * @param shapesGraphURI the URI of the shapes graph (must be in the dataset)
+  * @param shapesGraph    the ShapesGraph with the shapes to validate against
+  */
+// $COVERAGE-OFF$
+@SuppressWarnings(Array("NullParameter"))
+final class ShaclEngine private (dataset: Dataset, shapesGraphURI: URI, shapesGraph: ShapesGraph)
+    extends ValidationEngine(dataset, shapesGraphURI, shapesGraph, null) {
+  private var targetedNodes = 0
+
+  override def validateNodesAgainstConstraint(focusNodes: util.Collection[RDFNode], constraint: Constraint): Unit = {
+    super.validateNodesAgainstConstraint(focusNodes, constraint)
+    targetedNodes += 1
+  }
+
+  override def validateAll(): Resource = {
+    val r = super.validateAll()
+    if (r != null) r.addLiteral(toProperty(nxsh.targetedNodes), JenaDatatypes.createInteger(targetedNodes)) else r
+  }
+
+  private def toProperty(iri: Iri): Property =
+    ResourceFactory.createProperty(iri.toString)
+}
+
+object ShaclEngine {
+
+  private val shaclModel = SHACLSystemModel.getSHACLModel
+
+  /**
+    * Validates a given data Model against the SHACL shapes spec.
+    *
+    * @param shapesModel   the shapes Model to test against the SHACL shapes spec
+    * @param reportDetails true to also include the sh:detail (more verbose) and false to omit them
+    * @return an option of [[ValidationReport]] with the validation results
+    */
+  def apply(shapesModel: Model, reportDetails: Boolean): IO[String, ValidationReport] =
+    applySkipShapesCheck(shapesModel, shaclModel, validateShapes = true, reportDetails = reportDetails)
+
+  /**
+    * Validates a given data Model against all shapes from a given shapes Model.
+    *
+    * @param dataModel      the data Model
+    * @param shapesModel    the shapes Model
+    * @param validateShapes true to also validate any shapes in the data Model (false is faster)
+    * @param reportDetails  true to also include the sh:detail (more verbose) and false to omit them
+    * @return an option of [[ValidationReport]] with the validation results
+    */
+  def apply(
+      dataModel: Model,
+      shapesModel: Model,
+      validateShapes: Boolean,
+      reportDetails: Boolean
+  ): IO[String, ValidationReport] = {
+
+    val finalShapesModel = ValidationUtil.ensureToshTriplesExist(shapesModel)
+    // Make sure all sh:Functions are registered
+    SHACLFunctions.registerFunctions(finalShapesModel)
+    applySkipShapesCheck(dataModel, finalShapesModel, validateShapes, reportDetails)
+  }
+
+  private def applySkipShapesCheck(
+      dataModel: Model,
+      finalShapesModel: Model,
+      validateShapes: Boolean,
+      reportDetails: Boolean
+  ): IO[String, ValidationReport] = {
```

Comment:
I think it will be worth it to apply a timeout here to avoid cases where submitting a large model will yield a lot of errors. The timeout could also be handled in the application code (service).

Reply:
I postpone this for later. And I think it is not the responsibility of the `ShaclEngine` to handle the timeout, but of the routes or the `Schemas`

Reply:
agreed